### ISSUE TITLE
Sugar NPM Packages

### ISSUE THREAD

**Issue by andrewplummer** [andrewplummer/Sugar#522]
**Edit: "package" below refers to npm packages. "module" refers to Sugar's array, date, etc (actually also called "package" in the source)**

One of the major updates in Sugar v2.0.0 will be better NPM package handling. There are a few proposals here. The first two are pretty much done, but would appreciate opinions about the rest:

### Splitting out packages

Right now Sugar is a single package. I would like to split this out so that for example `sugar-array`, `sugar-date`, etc, would each be their own package. The current `sugar` package would remain as all default packages built into one.

### Core package

As of 2.0.0, Sugar will no longer extend native objects by default. Separating the act of defining methods and extending them is a key component to allowing plugins and safer augmentation to happen. The `sugar-core` package will expose only this functionality, allowing plugin developers to define their own methods that can later be extended onto prototypes by the end user.

### Sugar eats its own dogfood?

A natural extension of having the core package split out is that all Sugar packages could (should?) themselves require the `sugar-core` package. In addition to making sense, this would save a couple bytes if you are using sugar packages alongside independent plugins. I'm looking into doing this now, however there is one major consequence. If you are installing Sugar from npm, this means that you cannot load the root `sugar.js` file into a browser out of the box as it will have `require('sugar-core')` inside of it. Instead you would have to use it as part of browserify or some other build system. There are other things that can be done such as have a core-included version in a `dist/` folder or something, but I wonder if that wouldn't be confusing... Opinions here would be appreciated.

### Including separated modules in the main package?

One idea that came up would be to have separate modules (array, date, etc) in the main npm package, so if you wanted `sugar-array` and `sugar-date`, they would be available by doing `require('sugar/array'); require('sugar/date');`. Lodash does this. I'm not sure that this is feasible, but it's worth looking into. The advantage is that you wouldn't have 2 separate dependencies, however the disadvantage is that you're pulling down more than you need when you do `npm install sugar`. My instinct is to use npm as intended and keep separate modules in their own npm package, but I'm open to opinions here as well.

### Build package.json?

Depending on the answers to the above, it might be necessary to build the main `sugar` package using  gulp, however that means that `package.json` would no longer be checked into the root of the repo. Was hoping someone knows reasons that this may/may not be a good idea.

### Minified versions?

Sugar uses heavily optimized minification so it would probably make sense to include the minified files into the package as well. If idea 4 happens this would increase the file size as well, so it's something to consider.

### Better global/exports?

If there is robust, standardized pattern for handling a single global/exports in npm packages please let me know. What I have now works but I'd like it to be be more robust if possible.

**Comment by vendethiel** [comment]
1) split => nice
2) the main reason I use sugar is because it does extend, but that's just easy enough if it's opt-in anyway. Doesn't change anything for me, and makes it usable for a broader range of people. nice!
3) You might need to "compile" a big file with a require() shipped along. I think some libs do that? I think Chaplin does.
4) Well, this'd conflict with the dogfooding one, right?
5) I'm not sure what you mean?
6) :+1: but mostly because of my weird use case :).
7) (no clue)

**Comment by andrewplummer** [comment]
@vendethiel Thanks for the reply (and also for your continued support and responses... really appreciate it, even if I'm often disagreeable ;) )

Glad you like most ideas...

3) Will take a look at that one.
4) I need to look into it more, but I think it shouldn't have to. Let's say that the `sugar` package has `sugar-core` as a dependency. Then in `array.js` in the root it would be something like:

```
var Sugar = require('sugar-core');
Sugar.Array.defineMethods({ ... adding methods here ... });
module.exports = Sugar;
```

Essentially it's simply requiring the main exported object (would be the `Sugar` global if it was the browser), defining methods, and passing it along. Other packages would do the same thing. So in your project you could either do this:

```
require('sugar/date');
var Sugar = require('sugar/array');
Sugar.extendAll();
```

or this:

```
require('sugar/date')();
require('sugar/array')();
```

The `()` are shortcuts for each class's prototype extending method. If you don't want to extend prototypes, then you would have to use option #1 and access methods via `Sugar.Date.create`, etc.
Note that as far as I can see there is no difference between having different packages separated but in the main package and having each package separate, as you still have to require them in separately.

The only thing about 4) that might be annoying is that there would be a lot of files. Worst case it would be 1 dev/min for each package that requires `sugar-core` (for node/build system users), 1 dev/min for each package that has the core built in (if it needs to work out of the box in the browser), and 1 dev/min for each with all packages built together. So something like this:

```
// packages that require sugar-core
sugar.js // all packages built together
sugar.min.js
array.js
array.min.js
date.js
date.min.js
// ... etc

// packages with sugar-core pre-built in
dist/sugar.js
dist/sugar.min.js
dist/array.js
dist/array.min.js
dist/date.js
dist/date.min.js
// ... etc
```

Right now there are 12 packages, so it would be a total of 52 files. That's setting aside date locales as well so call it something like 70-80 files potentially. Honestly I don't know if this is offputting for people or not. Maybe having more options is better in the end than sparing file size when it comes to npm packages?? It's not like any of this reflects a bigger output size for end-users as people will be either picking and choosing which files to use or building them together with browserify etc.

Bottom line I guess we're looking at a single npm package that could potentially be as high as 1-2MB but with lots more options/freedom of use (use packages separately, use them out of the box, mix and match, etc), or a package that's more around 500KB or less, but is basically only the main build of all sugar packages built together. Considering the package size doesn't reflect what will actually be loaded into memory or downloaded by the user, is 1-2MB really all that big? Instinct tells me it would probably take more time to install 2 small dependencies than 1 large one... Anyway a lot of questions I'm not sure the answers to...

5) Sorry, I'm being confusing here... it's hard to explain what I mean. I was thinking that somehow if the npm package has a totally different structure to the github repo, then I would probably build that structure with a build tool like gulp. If that's the case then I got it into my head that I shouldn't have a `package.json` in the github repo anymore... But now that I think about it, that might not make any sense...

**Comment by vendethiel** [comment]
> @vendethiel Thanks for the reply (and also for your continued support and responses... really appreciate it, even if I'm often disagreeable ;) )

I wouldn't disagreeable at all. Maybe awol at times :P. I admit I stopped writing javascript pretty much 2 years ago, but I still like to follow the projects I used/that had nice ideas.

As someone who bundles up sugar into my browser extension, I'd really like to opt-out from translations. my plugins works for english/french, I don't need all the sugar ones.

At the end, if there was a sacrifice to make, I think I could easily live without minification built-in.

5) no, I think it still makes sense to have a package.json. Also it seems people are moving off bower

**Comment by andrewplummer** [comment]
> Maybe awol at times :p

Too true... :(

>  I'd really like to opt-out from translations

Well translations (locales) are always opt-in anyway. That's the case even now. Unless you mean you don't want them in the npm package at all (for filesize)?

**Comment by vendethiel** [comment]
They might belong to a separate npm package, I think it'd make sense.

**Comment by andrewplummer** [comment]
The plan is to include them in any npm package that includes the date module. By "include" I just mean the files though. They are still opt-in to use and won't be built into any of the "main" scripts. I would like to avoid a situation where locales require a separate dependency... when someone gets the date package (or a larger package), the locales should be "available but not built in".

If the reason is simply bloat, then 4) is going to end up being a lot worse because there will be a lot of separated packages that you won't end up using. The packages are larger as well.

**Comment by vendethiel** [comment]
> If the reason not to include them is simply bloat, then 4) is going to end up being a lot worse because there will be a lot of separated packages that you won't end up using. The packages are much larger as well.

Depends what kind of scheme you want to use. say, Rails has english by default, and a `rails-i18n` with every other language. if I can `require('sugar-i18n/lang/fr')`, I think I'd be happy enough ;-)

**Comment by andrewplummer** [comment]
But you are suggesting that it be a separate npm package altogether, right? The thing about Sugar is that, unlike rails (making some assumptions here) the locale logic is mostly baked into the date library itself. Of course you could make the case that *that* should be moved into a `sugar-i18n` package, but the fact is that it's pretty inextricable, as English is itself a locale. Not only that but it's also the most complicated locale, using almost every feature that locales offer, so any attempt to split it out would not only cause duplication for people using other languages, but there really wouldn't be all that much savings to be had.

So given all that, the only thing that `sugar-i18n` would offer is the locale definition files themselves (right now about 16 files < 2K each). Compared to the other packages we're talking about here that's kind of a drop in the hat. Even if the split-out sugar packages were not included in the main package I would still probably want to keep the locales, as the date module is included in the main Sugar package.

**Comment by andrewplummer** [comment]
FYI I went back and updated all the comments to refer to date, array, etc as "module" and "package" to only refer to npm packages as I realized it's pretty damn confusing otherwise.

**Comment by andrewplummer** [comment]
Ok, to update here everything is mostly finalized. I've taken a lot of opinions into account, and made some hard decisions, also put a lot of work in.

tl;dr The main Sugar npm package will be split out, by both module but also by method. This means that you will be able to require exactly the methods you want, across packages, and build together just what you need.

This does increase the filesize somewhat (although not that much). Some people might have complaints about this, but to me this was the last major complaint against Sugar in my mind. It's something people have been asking for and something I feel quite strongly about making dead simple to use.

It was also not easy to achieve. There is a massive gulp task now that parses out the entire codebase and builds up a dependency tree. Each method is exported into its own npm package (file) and "modules" have their own file as well to include all methods across the module. Usage is as follows:

```
// Cherry-picking methods
var Sugar = require('sugar-core');
require('sugar/array/flatten');
require('sugar/date/create');

Sugar.Array.flatten(...);
Sugar.Date.create(...);
```

```
// Requiring entire modules
var Sugar = require('sugar/array');
Sugar.Array.flatten(...);
```

Furthermore, extending natives is dead simple as well:

```
// Entire module
require('sugar/array')();
[1,2,3].flatten();
```

```
// Cherry-picked
require('sugar-core')();
require('sugar/array/flatten');
require('sugar/date/create');

[1,2,3].flatten();
Date.create(...);
```

The first `()` (this can also be explicitly called with `extendAll`) will extend all methods that have been defined up to that point and also any future methods that are defined, which is why the subsequent requires also have their methods extended.

The goal here is to get this working not only in node environments but also alongside packagers such as browserify as well so that custom packages can be made the npm way, right out of the box.

So, to answer the questions raised above:


### Splitting out packages

Yes, there will be multiple npm packages. The main `Sugar` package will have everything available, however requiring the "main" script (ie. `require('sugar')`) will include the default packages only (this way out of the box browserify builds should be the same size as the default download on the site).

Other npm packages will be by module, so if you are only interested in date methods you will require `sugar-date`, which is significantly smaller than the `sugar` package (methods are still split out so they can be required individually).

### Core package

As mentioned above there will be a `sugar-core` package. It will be the basic framework with no method definitions, allowing plugin developers to define methods that can later be extended by the end user.

### Sugar eats its own dogfood?

Yes, Sugar will behave in exactly the same way as plugins would. That means that `sugar-core` is NOT included in the `sugar` package. It is a dependency that will be pulled down on install. To answer the problem defined above of "you cannot load the root sugar.js file into a browser out of the box", there will be a `dist/` folder, like many other packages that will have pre-built versions that do include the core, so they can be loaded into a browser out of the box.

### Including separated modules in the main package?

Yes, as a result of the above you will not only be able to do `require('sugar/array')`, `require('sugar/date')`, etc, but also `require('sugar/array/flatten')` as well.

### Build package.json?

Nevermind this one... didn't really make sense to begin with. npm modules will be built, but will just pull the package.json in.

### Minified versions?

Yes, minified files will be included in the `dist/` folder as many other packages do.

### Better global/exports?

I'm not so concerned with this anymore as the npm modules will be built via a task and not mixed into the source code, so I don't have to be so clever. The core module is one exception here. I'm more or less just checking for `module` and `exports`, however. Still open to advice here if there is any, but should be OK.

**Comment by vendethiel** [comment]
Looks good! :+1:

**Comment by AndrewPoyntz** [comment]
bit late to the party, but were scoped packages considered? like @sugar/array

**Comment by andrewplummer** [comment]
@AndrewPoyntz Yes! [sugar-array](https://www.npmjs.com/package/sugar-array) now exists as it's own npm package. Additionally, using the main npm package you can do this:

```
var Sugar = require('sugar/array');
```

This will include only the Array module and it's dependencies. If you're using a package builder like browserify this will mean that your build will be much smaller.

**Comment by andrewplummer** [comment]
Ok, this is a bit out of order but Sugar 2.0.0 is now out so this issue should be addressed!
